compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,275 @@
+            currentNode->getGlobalIndex(),
+            currentNode->getParentGlobalIndex(),
+            _nodes->isSet(index + 1) ? "INLINED" : "NOT inlined",
+            currentNode->getByteCodeIndex(),
+            currentNode->getByteCodeSize(),
+            currentNode->getName(comp->trMemory()),
+            currentNode->getBenefit(),
+            currentNode->getCost(),
+            currentNode->getBudget()
+         );
+
+         if (traceBIProposal)
+            traceMsg(comp, "%s\n", line.text());
+         if (verboseInlining)
+            TR_VerboseLog::writeLine(TR_Vlog_BI, "%s", line.text());
+         }
+
+      //process children
+      const uint32_t numChildren = currentNode->getNumChildren();
+
+      for (uint32_t i = 0; i < numChildren; i++)
+         idtNodeQueue.push_back(currentNode->getChild(i));
+
+      }
+   if (traceBIProposal)
+      traceMsg(comp, "\n");
+   }
+
+
+void TR::InliningProposal::addNode(TR::IDTNode *node)
+   {
+   ensureBitVectorInitialized();
+
+   const int32_t index = node->getGlobalIndex() + 1;
+   if (_nodes->isSet(index))
+      {
+      return;
+      }
+
+   _nodes->set(index);
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::isEmpty()
+   {
+   if (!_nodes)
+      return true;
+
+   return _nodes->isEmpty();
+   }
+
+uint32_t TR::InliningProposal::getCost()
+   {
+   if (_cost == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _cost;
+   }
+
+double TR::InliningProposal::getBenefit()
+   {
+   if (_benefit == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _benefit;
+   }
+
+void TR::InliningProposal::computeCostAndBenefit()
+   {
+   _cost = 0;
+   _benefit = 0;
+
+   if (!_idt)
+      return;
+
+   TR_BitVectorIterator bvi(*_nodes);
+   int32_t idtNodeIndex;
+
+   while (bvi.hasMoreElements())
+      {
+      idtNodeIndex = bvi.getNextElement();
+      IDTNode *node = _idt->getNodeByGlobalIndex(idtNodeIndex - 1);
+      _cost += node->getCost();
+      _benefit += node->getBenefit();
+      }
+   }
+
+void TR::InliningProposal::ensureBitVectorInitialized()
+   {
+   if (!_nodes)
+      _nodes = new (_region) TR_BitVector(_region);
+   }
+
+bool TR::InliningProposal::isNodeInProposal(TR::IDTNode* node)
+   {
+   if (_nodes == NULL)
+      return false;
+   if (_nodes->isEmpty())
+      return false;
+
+   const int32_t idx = node->getGlobalIndex() + 1;
+
+   return _nodes->isSet(idx);
+   }
+
+void TR::InliningProposal::merge(TR::InliningProposal *a, TR::InliningProposal *b)
+   {
+   if (a->_nodes && b->_nodes)
+      {
+      TR_BitVector temp(*a->_nodes);
+      temp |= *b->_nodes;
+      ensureBitVectorInitialized();
+      *_nodes = temp;
+      }
+   else if (a->_nodes)
+      {
+      ensureBitVectorInitialized();
+      *_nodes = *a->_nodes;
+      }
+   else if (b->_nodes)
+      {
+      ensureBitVectorInitialized();
+      *_nodes = *b->_nodes;
+      }
```

Comment:
When `a->nodes` and `b->nodes` are both null, this logic will currently leave `this` unmodified (apart from invalidating `_cost` and `_benefit`). I think in that case `this` should probably be cleared out instead, so that it contains the union of `a` and `b` even if `this` was previously nonempty, e.g.
```
else if (_nodes)
   {
   _nodes->empty();
   }

Reply:
Yes, I forgot to clear it out. I added it now and also added a comment for it.